source/extensions/common/wasm/wasm_vm.cc

```diff
@@ -78,7 +78,9 @@ bool isWasmEngineAvailable(absl::string_view runtime) {
 }
 
 absl::string_view getFirstAvailableWasmEngineName() {
-  static const std::vector<absl::string_view> wasm_engines{"envoy.wasm.runtime.v8", "envoy.wasm.runtime.wasmtime", "envoy.wasm.runtime.wamr", "envoy.wasm.runtime.wavm"};
+  static const std::vector<absl::string_view> wasm_engines{
```

Comment:
constexpr. statically constructed non-pods are not allowed, but if you can declare it constexpr it's OK.

or you could make it non-static.

Reply:
static constructed non-prod are allowed in methods as the initialization order isn't an issue. this cannot be constexpr.

Reply:
Sorry; I meant this *can* be constexpr.

Reply:
Hm. I thought you could make constexpr vector but could not find any in the envoy codebase. You don't need a vector though; you can use `constexpr std::array` or simply `constexpr absl::string_view[] ...`.

This is marginally better than `const` IMO as it can be constructed at compile-time.

Reply:
Also I see why Lizan says non-pod static const doesn't have an init-order issue, but my reading of https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables is that they are still recommended against, and they are not needed here.

Reply:
Do we have a consensus on this? Personally I think since the value is known at compile time, static const won't be an issue here.

Reply:
It violates the style guide, although in this case I agree it's safe.

But it forces the readers to study and convince themselves it's safe. So the style guide is partially to help us write code that's more obviously correct, and thus easier to read.

So I vote to use a C array or std::array of absl::string_view which you can construct as a constexpr which is obviously safe. I believe it will be faster and smaller as well.

Reply:
Yes, I agree using a constexpr will be faster in run time and follows the style guide as well. Not really understand why lizan said can not use constexpr here.

Reply:
You can't make a constexpr std::vector because it heap-allocates its data. But you can do that with a std::array or C array since the data is inline and can be organized by the compiler and placed into the object file directly.

Reply:
oh I see what he was saying now. I thought he just opposed to `constexpr` in general here, and that makes me confused.